README.md

# teams-simpleauth

A node.js Microsoft Teams simpleauth implementation that works with the [@microsoft/teamsfx](https://www.npmjs.com/package/@microsoft/teamsfx) client side component.

## Installation 

Install with npm

```bash 
  npm install teams-simpleauth --save
```
    
## Documentation

### Prerequisites

A Teams Tab project using [Express.js](http://expressjs.com/) as a server side implementation, for instance a project scaffolded with [Yo Teams](https://aka.ms/yoteams)

### Installation and configuration

Start by installing the `teams-simpleauth` and `@microsoft/teamsfx` packages. The first one is used for the server side *exchange* of the token and the latter one is for the client side scripting.

``` bash
npm install teams-simpleauth @microsoft/teamsfx --save
```

Next, in your server side component add the following import. For *yo teams* based applications this is in the `src/server/server.ts` file.

``` Typescript
import { SimpleAuthRouter } from "teams-simpleauth";
```

And then after the express application is create add a statement like this to add and configure the router.

`appId`, `appIdUri` and `appSecret` are retrieved from the AAD application that is used for the Teams SSO Tab. `scopes` is an array of delegated permission scopes you would like for the returned access token and finally `tenantId` is either the ID of the tenant or set to `common` for multi-tenant applications.

``` Typescript
express.use("/auth/token", SimpleAuthRouter({
    appId: process.env.TAB_APP_ID as string,
    appIdUri: process.env.TAB_APP_URI as string,
    appSecret: process.env.TAB_APP_SECRET as string,
    scopes: ["Presence.Read"],
    tenantId: "00000000-0000-0000-0000-000000000000"
}));
```

In the client side code (tab) you need to import the `teamsfx` package as follows, and you use the package as described in the [`teamsfx` documentation](https://www.npmjs.com/package/@microsoft/teamsfx).

``` Typescript
import * as teamsFx from "@microsoft/teamsfx";
```


Example of tab using the `teamsfx` component in a *yo teams* based tab:

``` TypeScript
teamsFx.loadConfiguration({
    authentication: {
        initiateLoginEndpoint: `https://${process.env.PUBLIC_HOSTNAME}/ile`,
        clientId: process.env.TAB_APP_ID,
        tenantId: "00000000-0000-0000-0000-000000000000",
        authorityHost: "https://login.microsoftonline.com",
        applicationIdUri: process.env.TAB_APP_URI,
        simpleAuthEndpoint: `https://${process.env.PUBLIC_HOSTNAME}`
    }
});
const credential = new teamsFx.TeamsUserCredential();
const graphClient = teamsFx.createMicrosoftGraphClient(credential, ["Presence.Read"]);
const result = await graphClient.api("/me/presence").get();
setPresence(" is " + result.availability);
```
  
## License

[MIT](https://choosealicense.com/licenses/mit/)

  


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [*1.0.0*]- <*2021-06-16*>

### Added

* Initial release

# Template
## [*MAJOR.MINOR.PATCH*] - <*DATE*>
### Added
### Changed
### Fixed
### Removed

src/index.spec.ts
import { SimpleAuthRouter, SimpleAuthBearerStrategy } from "./index";

describe("SimpleAuthRouter", () => {
    it("Should export SimpleAuthRouter", () => {
        expect(SimpleAuthRouter).toBeInstanceOf(Function);
    });
    it("Should export SimpleAuthBearerStrategy", () => {
        expect(SimpleAuthBearerStrategy).toBeInstanceOf(Function);
    });
})


src/SimpleAuthRouter.spec.ts
import express from "express";
import { stub } from "jest-auto-stub";
import request from "supertest";
import { ISimpleAuthRouterOptions } from "./ISimpleAuthRouterOptions";
import { SimpleAuthRouter } from "./SimpleAuthRouter";
import { AuthenticationResult, ConfidentialClientApplication } from "@azure/msal-node";

const acquireTokenOnBehalfOfMock = jest.fn().mockResolvedValue({});
let authenticateResultMock: any = null;

jest.mock("@azure/msal-node");
jest.mock("passport", () => {
    return {
        Passport: jest.fn().mockImplementation(() => {
            return {
                initialize: jest.fn().mockReturnValue((_req: any, _res: any, next: () => void) => { next(); }),
                use: jest.fn().mockReturnValue({}),
                authenticate: jest.fn().mockReturnValue((_req: any, _res: any, next: (a: any, b: any, c: any) => void) => { next(authenticateResultMock, {}, "asdf"); })
            }
        }),
        Strategy: jest.fn()
    }
});


describe("SimpleAuthRouter", () => {
    let app: express.Express;
    beforeEach(() => {
        jest.clearAllMocks();
        authenticateResultMock = null;
        (ConfidentialClientApplication as any).mockImplementation(() => {
            return {
                acquireTokenOnBehalfOf: acquireTokenOnBehalfOfMock
            }
        });

        app = express();
        app.use("/auth/token", SimpleAuthRouter({
            appId: "4cb0e766-c9f1-48ed-8304-6d576cca7670",
            appIdUri: "app://4cb0e766-c9f1-48ed-8304-6d576cca7670",
            appSecret: "secret",
            scopes: ["scope"],
            tenantId: "common"
        }));
    });

    it("Should return a 404 for a GET request", async () => {
        const response = await request(app).get("/auth/token").send();
        expect(response).toBeDefined();
        expect(response.statusCode).toEqual(404);
    });

    it("Should return a 401 for a POST request without auth header", async () => {
        const response = await request(app).post("/auth/token").send({});
        expect(response).toBeDefined();
        expect(acquireTokenOnBehalfOfMock).toBeCalledTimes(0);
        expect(response.statusCode).toEqual(401);
    });

    it("Should return a 500 for an invalid bearer token", async () => {
        authenticateResultMock = new Error();
        const response = await request(app).post("/auth/token").set("authorization", "Bearer dummy").send({});
        expect(response).toBeDefined();
        expect(acquireTokenOnBehalfOfMock).toBeCalledTimes(0);
        expect(response.statusCode).toEqual(500);
    });

    it("Should return a 500 for an invalid OBO call", async () => {
        acquireTokenOnBehalfOfMock.mockRejectedValueOnce("error");
        const response = await request(app).post("/auth/token").set("authorization", "Bearer dummy").send({});
        expect(response).toBeDefined();
        expect(acquireTokenOnBehalfOfMock).toBeCalledTimes(1);
        expect(response.statusCode).toEqual(500);
    });

    it("Should return a 401 for a POST request with invalid auth header", async () => {
        acquireTokenOnBehalfOfMock.mockResolvedValueOnce(null);
        const response = await request(app).post("/auth/token").set("authorization", "Bearer dummy").send({});
        expect(response).toBeDefined();
        expect(acquireTokenOnBehalfOfMock).toBeCalledTimes(1);
        expect(response.statusCode).toEqual(401);
    });

    it("Should return a 200 for a POST request with a valid auth header", async () => {
        acquireTokenOnBehalfOfMock.mockResolvedValueOnce(stub<AuthenticationResult>({ accessToken: "token", scopes: ["scope"], expiresOn: new Date(0) }));
        const response = await request(app).post("/auth/token").set("authorization", "Bearer dummy").send({});
        expect(response).toBeDefined();
        expect(acquireTokenOnBehalfOfMock).toBeCalledTimes(1);
        expect(response.statusCode).toEqual(200);
        expect(response.body).toEqual({ access_token: "token", scope: ["scope"], expires_on: "1970-01-01T00:00:00.000Z" })
    });
});


src/simpleAuthRouter.ts
// Copyright (c) Wictor Wilén. All rights reserved.
// Licensed under the MIT license.
import { Router } from "express";
import { Passport } from "passport";
import { BearerStrategy, IBearerStrategyOption, ITokenPayload, VerifyCallback } from "passport-azure-ad";
import { ConfidentialClientApplication } from "@azure/msal-node";
import { ISimpleAuthRouterOptions } from "./ISimpleAuthRouterOptions";

/**
 * Creates the Teams Simpleauth router
 * @param options options for the router
 * @returns an Express router
 * @example
 * express.use("/auth/token", simpleAuthRouter({
 *   appId: process.env.TAB_APP_ID as string,
 *   appIdUri: process.env.TAB_APP_URI as string,
 *   appSecret: process.env.TAB_APP_SECRET as string,
 *   scopes: ["Presence.Read"],
 *   tenantId: "aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee"
* }));
 */
export const simpleAuthRouter = (options: ISimpleAuthRouterOptions): Router => {
    const router = Router();

    // Set up the Bearer Strategy
    const bearerStrategy = new BearerStrategy({
        identityMetadata: "https://login.microsoftonline.com/common/v2.0/.well-known/openid-configuration",
        clientID: options.appId,
        audience: options.appIdUri,
        loggingLevel: "warn",
        validateIssuer: false,
        passReqToCallback: false
    } as IBearerStrategyOption, (token: ITokenPayload, done: VerifyCallback) => {
        done(null, { tid: token.tid, name: token.name, upn: token.upn }, token);
    }
    );

    const pass = new Passport();
    router.use(pass.initialize());
    pass.use(bearerStrategy);

    const confidentialClient = new ConfidentialClientApplication({
        auth: {
            clientId: options.appId,
            clientSecret: options.appSecret,
            authority: "https://login.microsoftonline.com/" + options.tenantId
        }
    });

    router.post("",
        pass.authenticate("oauth-bearer", { session: false }),
        async (req, res, next) => {
            const authHeader = req.headers.authorization;

            if (authHeader === undefined) {
                res.status(401).send();
                return;
            }

            confidentialClient.acquireTokenOnBehalfOf({
                oboAssertion: authHeader.split(" ")[1],
                scopes: options.scopes
            }).then(response => {
                if (response !== null) {
                    res.send(
                        {
                            access_token: response.accessToken,
                            scope: response.scopes,
                            expires_on: response.expiresOn
                        });
                } else {
                    res.status(401).send();
                }

            }).catch(err => {
                res.status(500).send(err);
            });
        });

    return router;
};


src/SimpleAuthBearerStrategy.ts
// Copyright (c) Wictor Wilén. All rights reserved.
// Licensed under the MIT license.

import { BearerStrategy, IBearerStrategyOption, ITokenPayload, VerifyCallback } from "passport-azure-ad";

/**
 * Creates a Passport Bearer strategy
 * @param clientID client id
 * @param audience audience
 * @param loggingLevel logging level
 * @returns a Passport Bearer strategy
 */
export const SimpleAuthBearerStrategy = (clientID: string, audience: string, loggingLevel = "warn") => new BearerStrategy({
    identityMetadata: "https://login.microsoftonline.com/common/v2.0/.well-known/openid-configuration",
    clientID,
    audience,
    loggingLevel,
    validateIssuer: false,
    passReqToCallback: false
} as IBearerStrategyOption, (token: ITokenPayload, done: VerifyCallback) => {
    done(null, { tid: token.tid, name: token.name, upn: token.upn }, token);
}
);


src/SimpleAuthRouter.ts
// Copyright (c) Wictor Wilén. All rights reserved.
// Licensed under the MIT license.

import { Router } from "express";
import { Passport } from "passport";
import { ConfidentialClientApplication } from "@azure/msal-node";
import { ISimpleAuthRouterOptions } from "./ISimpleAuthRouterOptions";
import { SimpleAuthBearerStrategy } from "./SimpleAuthBearerStrategy";
import debug from "debug";

const log = debug("simpleauth");

/**
 * Creates the Teams Simpleauth router
 * @param options options for the router
 * @returns an Express router
 * @example
 * express.use("/auth/token", SimpleAuthRouter({
 *   appId: process.env.TAB_APP_ID as string,
 *   appIdUri: process.env.TAB_APP_URI as string,
 *   appSecret: process.env.TAB_APP_SECRET as string,
 *   scopes: ["Presence.Read"],
 *   tenantId: "aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee"
* }));
 */
export const SimpleAuthRouter = (options: ISimpleAuthRouterOptions): Router => {
    const router = Router();

    const pass = new Passport();
    router.use(pass.initialize());
    pass.use(SimpleAuthBearerStrategy(options.appId, options.appIdUri));

    const confidentialClient = new ConfidentialClientApplication({
        auth: {
            clientId: options.appId,
            clientSecret: options.appSecret,
            authority: "https://login.microsoftonline.com/" + options.tenantId
        }
    });

    router.post("",
        pass.authenticate("oauth-bearer", { session: false }),
        async (req, res, next) => {
            const authHeader = req.headers.authorization;

            if (authHeader === undefined) {
                log("Missing authorization header");
                res.status(401).send();
                return;
            }

            confidentialClient.acquireTokenOnBehalfOf({
                oboAssertion: authHeader.split(" ")[1],
                scopes: options.scopes
            }).then(response => {
                if (response !== null) {
                    log("Successfully retrieved access token");
                    res.send(
                        {
                            access_token: response.accessToken,
                            scope: response.scopes,
                            expires_on: response.expiresOn
                        });
                } else {
                    log("No response from on-behalf-of request");
                    res.status(401).send();
                }

            }).catch(err => {
                log(`Error when requesting access token: ${err}`);
                res.status(500).send(err);
            });
        });

    return router;
};
